runtime/etc/mli/mli_common_code.c

```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright 2004-2019 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CHPL_RUNTIME_ETC_MLI_MLI_COMMON_CODE_C_
+#define CHPL_RUNTIME_ETC_MLI_MLI_COMMON_CODE_C_
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <stdint.h>
+#include <string.h>
+#include <stdarg.h>
+#include <assert.h>
+#include <zmq.h>
+
+//
+// Hopeful workaround to prevent Travis from complaining. We may just have to
+// force the client to also use the Chapel runtime as well, for better or
+// for worse.
+//
+#ifdef CHPL_MLI_IS_CLIENT_
+# define mli_malloc(bytes) chpl_malloc(bytes)
+# define mli_free(ptr) chpl_free(ptr)
+#elif defined(CHPL_MLI_IS_SERVER_)
+# define mli_malloc(bytes) malloc(bytes)
+# define mli_free(ptr) free(ptr)
+# else
+# error The mli_malloc/free macros were defined in a inappropriate place.
+# endif
+
+
+// Error codes must all be less than zero unless we change the protocol!
+enum chpl_mli_errors {
+
+  CHPL_MLI_ERROR_NONE       = +0,
+  CHPL_MLI_ERROR_SHUTDOWN   = -1,
+  CHPL_MLI_ERROR_UNKNOWN    = -2,
+  CHPL_MLI_ERROR_NOFUNC     = -3,
+  CHPL_MLI_ERROR_SOCKET     = -4,
+  CHPL_MLI_ERROR_EXCEPT     = -5
+
+};
+
+const char* chpl_mli_errstr(enum chpl_mli_errors e) {
+  static const char* mli_errors_[] = {
+    "NONE",
+    "SHUTDOWN",
+    "UNKNOWN",
+    "NOFUNC",
+    "SOCKET",
+    "EXCEPT"
+  };
+
+  if (e > CHPL_MLI_ERROR_NONE || e < CHPL_MLI_ERROR_EXCEPT) {
+    return "INVALID_ERROR_CODE";
+  }
+
+  return mli_errors_[e];
+}
+
+//
+// Both the client and server will be using this to perform RPC.
+//
+struct chpl_mli_context {
+
+  void* context;
+  void* main;
+  void* arg;
+  void* res;
+
+};
+
+static
+char* chpl_mli_concat(size_t count, ...) {
+  char* result = NULL;
+  char* bufptr = NULL;
+  va_list argp;
```

Comment:
For my own edification, what is va_list and va_start?

Reply:
Those are how you access varargs lists in C.